### PR TITLE
AWS VPC CNI integration

### PR DESCRIPTION
**Pull request by nerahou**

### CODE REVIEW

File: versions.tf [hunk 1]
```diff
@@ -34,5 +38,9 @@ terraform {
       source  = "hashicorp/tls"
       version = ">=3.4.0"
     }
+    time = {
+      source  = "hashicorp/time"
+      version = ">=0.11.1"
+    }
```
Comment (antonincms) [inline versions.tf:44]: Provider does not seem to be used

File: main.tf [hunk 2]
```diff
@@ -15,6 +15,12 @@
 */
 
 locals {
+
```
Comment (antonincms) [inline main.tf:18]:
```suggestion
```

File: main.tf [hunk 3]
```diff
@@ -60,6 +66,20 @@ data "aws_caller_identity" "current" {}
 # This datasource is used to get the region currently used by the AWS provider
 data "aws_region" "current" {}
 
+provider "helm" {
+
```
Comment (antonincms) [inline main.tf:70]:
```suggestion
```

**Review by antonincms — CHANGES_REQUESTED** [review]

File: values.yaml
Comment (vincentmrg) [inline values.yaml:1]: This file should probably be renamed and moved to a "templates" directory.

File: node_group_advanced.tf [hunk 4]
```diff
@@ -274,6 +286,10 @@ resource "aws_autoscaling_group" "quortex_asg_advanced" {
       },
       # the following tags must be set on the ASG, and must match the k8s node labels/taints, for the autoscaler to be able to scale up from 0
       { for k, v in lookup(each.value, "labels", {}) : "k8s.io/cluster-autoscaler/node-template/label/${k}" => v },
+      length(lookup(each.value, "subnet_ids", [])) == 1 ?
+      {
+        "k8s.io/cluster-autoscaler/node-template/label/k8s.amazonaws.com/eniConfig" : data.aws_subnet.subnets_advanced[each.key].availability_zone
+      } : {},
```
Comment (vincentmrg) [inline node_group_advanced.tf:292]: I think that this is not the label we expect.

File: main.tf [hunk 5]
```diff
@@ -159,6 +172,10 @@ resource "aws_eks_node_group" "quortex" {
       "k8s.io/cluster-autoscaler/node-template/label/nodegroup" = each.key, # tag required for scaling to/from 0
     } : {},
     { "nodegroup" = each.key },
+    length(lookup(each.value, "subnet_ids", [])) == 1 ?
+    {
+      "k8s.amazonaws.com/eniConfig" : data.aws_subnet.subnets[each.key].availability_zone
+    } : {},
```
Comment (vincentmrg) [inline main.tf:178]: You should probably check if your `handle_eni_configs` variable is `true` here.

File: main.tf [hunk 6]
```diff
@@ -168,6 +185,10 @@ resource "aws_eks_node_group" "quortex" {
     {
       "nodegroup" = each.key
     },
+    length(lookup(each.value, "subnet_ids", [])) == 1 ?
+    {
+      "k8s.amazonaws.com/eniConfig" : data.aws_subnet.subnets[each.key].availability_zone
+    } : {},
```
Comment (vincentmrg) [inline main.tf:191]: You should probably check if your `handle_eni_configs` variable is `true` here.

File: node_group_advanced.tf [hunk 7]
```diff
@@ -115,6 +123,10 @@ resource "aws_launch_template" "quortex_launch_tpl" {
                     "eks.amazonaws.com/nodegroup"       = each.key,
                     "nodegroup"                         = each.key
                   },
+                  length(lookup(each.value, "subnet_ids", [])) == 1 ?
+                  {
+                    "k8s.amazonaws.com/eniConfig" : data.aws_subnet.subnets_advanced[each.key].availability_zone
+                  } : {},
```
Comment (vincentmrg) [inline node_group_advanced.tf:129]: You should probably check if your `handle_eni_configs` variable is `true` here.

File: variables.tf [hunk 8]
```diff
@@ -167,6 +167,21 @@ variable "master_authorized_networks" {
...
+A map representing the pods subnets. Each item contains the subnet's ID,
+Availability Zone, cidr block, and whether the subnet is public or not.
+EOT
+  default     = {}
+}
+
+variable "handle_eni_configs" {
+  type        = bool
+  description = "To determine if eniconfig resources should be managed by this module"
+  default     = true
```
Comment (vincentmrg) [inline variables.tf:182]: By setting`false` as default value, we would be backward compatible.

File: node_group_advanced.tf [hunk 9]
```diff
@@ -35,6 +35,14 @@ locals {
   }
 }
 
+data "aws_subnet" "subnets_advanced" {
+  for_each = var.node_groups_advanced
+  filter {
+    name   = "subnet-id"
+    values = lookup(each.value, "subnet_ids", [])
+  }
+}
```
Comment (vincentmrg) [inline node_group_advanced.tf:44]: duplicate with the previous one

File: main.tf [hunk 10]
```diff
@@ -25,13 +25,18 @@ locals {
...
+      for k, v in lookup(node_group, "labels", { "k8s.amazonaws.com/eniConfig" : data.aws_subnet.subnets[local.nodes_subnet_ids_map[key]].availability_zone }) :
       key => {
         "k8s.io/cluster-autoscaler/node-template/label/${k}" : v
         (k) : v
       }...
     }
   ]
+  nodes_subnet_ids_map = merge(
+    { for key, node_group in var.node_groups : key => lookup(node_group, "subnet_ids", [])[0] },
+    { for key, node_group in var.node_groups_advanced : key => lookup(node_group, "subnet_ids", [])[0] },
```
Comment (vincentmrg) [inline main.tf:37]: What if there's node_groups and node_groups_advanced with the same key ?

Comment (vincentmrg) [reply main.tf:37]: There's also a clear risk of index out-of-bounds here.

File: main.tf [hunk 11]
```diff
@@ -15,18 +15,28 @@
...
+  )
+  distinct_subnet_ids = distinct([for key, subnet in local.nodes_subnet_ids_map : subnet])
```
Comment (vincentmrg) [inline main.tf:39]: A bit complicated for something that could have been written `toset(values(local.nodes_subnet_ids_map))`

**Review by vincentmrg — APPROVED** [review]

**Pull request merged**